File: Python/344.reverseString.py
# 递归 O(n) O(n)
class Solution:
    def reverseString(self, s: List[str]) -> None:
        def hepler(left, right):
          if (left < right):
            s[left], s[right] = s[right], s[left]
            hepler(left + 1, right - 1)
        hepler(0, len(s) - 1)

# 双指针 O(n) O(1)
class Solution:
    def reverseString(self, s: List[str]) -> None:
        left, right = 0, len(s) - 1
        while left < right:
          s[left], s[right] = s[right], s[left]
          left, right = left + 1, right - 1

File: Java/344.reverseString.java
// 递归 O(n) O(n)
class Solution {
  public void helper(char[] s, int left, int right) {
    if (left < right) {
      char tmp = s[left];
      s[left++] = s[right];
      s[right--] = tmp;
      helper(s, left, right);
    }
  }
  public void reverseString(char[] s) {
    helper(s, 0, s.length - 1);
  }
}

// 双指针 O(n) O(1)
class Solution {
  public void reverseString(char[] s) {
    int left = 0, right = s.length - 1;
    while (left < right) {
      char tmp = s[left];
      s[left++] = s[right];
      s[right--] = tmp;
    }
  }
}